File: leetcode49接雨水.py
class Solution():
    def trap(self, height):
        if not height:
            return 0
        rain = 0
        maxLeft = 0
        vector = []
        for h in height:
            maxLeft = max(h, maxLeft)
            vector.append(maxLeft)
        maxRight = 0
        for i in range(len(vector)-1, -1, -1):
            h = height[i]
            m = vector[i]
            maxRight = max(h, maxRight)
            dec = min(m, maxRight) - h
            if dec > 0:
                rain += dec
        return rain

File: leetcode49接雨水.cpp
#include<vector>
#include<list>
#include<stack>
using namespace std;
class Solution {
public:
	int trap(vector<int>& height) {
		if (height.size() == 0)
		{
			return 0;
		}
		int rain=0,maxLeft = 0,maxRight=0,accumul=0,m,i=0,h,minHeight;
		vector<int> s;
		for (;i<height.size();i++)
		{
			if (maxLeft < height[i])
			{
				maxLeft = height[i];
			}
			s.push_back(maxLeft);
		}
		i = s.size() - 1;
		for (;i>=0;i--)
		{
			m = s[i];
			h = height[i];
			if (maxRight < h)
			{
				maxRight= h;
			}
			if (maxRight > m)
			{
				accumul = m - h;
			}
			else
			{
				accumul = maxRight - h;
			}
			if (accumul>0)
			{
				rain += accumul;
			}
		}
		return rain;
	}
};
